internal/agent/device/config/hook.go

```diff
@@ -0,0 +1,433 @@
+				return
+			}
+			m.Handle(ctx, event)
+		case err, ok := <-m.fsMonitor.Errors():
+			if !ok {
+				m.log.Debug("Watcher errors channel closed")
+				return
+			}
+			m.log.Errorf("error: %v", err)
+		}
+	}
+
+}
+
+// Update the manager with the new hook if appropriate.
+func (m *hookManager) Update(hook *v1alpha1.DeviceConfigHookSpec) (bool, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if m.handlers == nil {
+		return false, ErrHookManagerNotInitialized
+	}
+
+	handler, ok := m.handlers[hook.WatchPath]
+	if !ok || !reflect.DeepEqual(hook, handler.DeviceConfigHookSpec) {
+		return true, addOrReplaceHookHandler(m.fsMonitor, hook, m.handlers)
+	}
+
+	return false, nil
+}
+
+func (m *hookManager) HandleErrors() []error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	var errs []error
+	for _, handler := range m.handlers {
+		if err := handler.Error(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return errs
+}
+
+func (m *hookManager) WatchList() []string {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	return m.fsMonitor.WatchList()
+}
+
+func (m *hookManager) WatchRemove(watchPath string) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if m.handlers == nil {
+		return ErrHookManagerNotInitialized
+	}
+
+	if _, ok := m.handlers[watchPath]; ok {
+		if err := m.fsMonitor.WatchRemove(watchPath); err != nil {
+			return fmt.Errorf("failed removing watch: %w", err)
+		}
+		delete(m.handlers, watchPath)
+	}
+	return nil
+}
+
+func (m *hookManager) Handle(ctx context.Context, event fsnotify.Event) {
+	filePath := event.Name
+	handle := m.getHandle(filePath)
+	if handle == nil {
+		// no handler for this event
+		return
+	}
+
+	actions, ok := handle.Actions(event.Op)
+	if !ok {
+		// handle does not have any actions for this file operation
+		return
+	}
+
+	// actions
+	var errs []error
+	for i := range actions {
+		action := actions[i]
+		hookActionType, err := action.Discriminator()
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		switch hookActionType {
+		case SystemdHookActionType:
+			hookAction, err := action.AsConfigHookActionSystemdSpec()
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			if err := handleHookActionSystemd(ctx, m.log, m.systemdClient, &hookAction, filePath); err != nil {
+				errs = append(errs, err)
+			}
+		case ExecutableHookActionType:
+			hookAction, err := action.AsConfigHookActionExecutableSpec()
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+
+			if err := handleHookActionExecutable(ctx, m.exec, &hookAction, filePath); err != nil {
+				errs = append(errs, err)
+			}
+		default:
+			errs = append(errs, fmt.Errorf("unknown hook action type: %s", hookActionType))
+		}
+	}
+
+	// push errors to status
+	handle.SetError(errors.Join(errs...))
+}
+
+func (m *hookManager) getHandle(eventName string) *HookHandler {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	// check if the event name is a file or directory
+	paths := []string{eventName, filepath.Dir(eventName)}
```

Comment:
What if we have watchers both on the event and the directory.  Is the directory hidden for this event?

Reply:
yeah, we will have 2 separate events I will clean this up and add tests

Reply:
added